build/xtask/src/dist.rs

```diff
@@ -2452,69 +2452,125 @@ fn write_elf(
+                (*candidate_section.0) + candidate_section.1.data.len() as u64;
 
-    let sections_base_address = kentry;
-    let sections_length = (*(last_section.0) as u64
-        + last_section.1.data.len() as u64)
-        - sections_base_address as u64;
+            if end > sections_length {
+                sections_length = end;
+            }
+        }
+    }
+    sections_length -= sections_base_address;
 
     let mut program_headers = Vec::new();
+    let mut section_headers = Vec::new();
+
+    // Create a Section Header String Table, to hold the actual section
+    // names.
+    let mut shstrtab = Vec::new();
+    shstrtab.push(0x00 as u8);
+
     // Preallocate a vector for section data, filled with 0xFF. This pattern is chosen
     // to replicate the erase pattern we'd find in flash, and match the padding value
     // previously chosen for the objcopy gap filler.
     let mut sections_data = vec![0xFF; sections_length.try_into().unwrap()];
-
+    let mut section_header_name_index = 0 as usize;
     // Generate all the program headers and collect all the sections together.
-    for (&base, sec) in sections {
+    for (base, sec) in sections {
         if sec.data.is_empty() {
             // Do not create a program header for an empty section. There's nothing to load.
             continue;
         }
 
+        let this_section_base_offset = base - sections_base_address;
+        let this_section_end_offset =
+            this_section_base_offset as usize + sec.data.len() as usize;
+
         program_headers.push(ProgramHeader {
             p_type: PT_LOAD,
             p_flags: PF_X | PF_W | PF_R,
-            p_offset: sections_data.len() as u64,
-            p_vaddr: base as u64,
-            p_paddr: base as u64,
-            p_filesz: sec.data.len() as u64,
-            p_memsz: sec.data.len() as u64,
+            p_offset: this_section_base_offset as u32,
+            p_vaddr: *base as u32,
+            p_paddr: *base as u32,
+            p_filesz: sec.data.len() as u32,
+            p_memsz: sec.data.len() as u32,
             p_align: 0x20, // This matches the alignment guarantees of the kernel & task build
         });
 
-        let this_section_base_offset = base - sections_base_address;
-        let this_section_end_offset =
-            this_section_base_offset as usize + sec.data.len() as usize;
+        section_headers.push(SectionHeader {
+            sh_type: SHT_PROGBITS,
+            sh_flags: (SHF_ALLOC | SHF_EXECINSTR) as u32,
+            sh_name: shstrtab.len() as u32,
+            sh_offset: this_section_base_offset as u32,
+            sh_size: sec.data.len() as u32,
+            sh_addr: kentry as u32,
+            sh_addralign: 0x20,
+            sh_entsize: 0, // No fixed-size entries here
+            sh_link: 0,
+            sh_info: 0,
+        });
+
+        let task_name = sec.source_file.file_name().to_owned().unwrap();
+
+        let mut section_name: String = String::from(".text.");
```

Comment:
Just noting: Even though this adds the prefix `.text` to every section, not every section is really .text. But probably, we don't have the information at this point to determine whether it is .text or something else.

Reply:
Exactly that - by the time we get the section data here, we've only really got knowledge of the bitstream, source file, and offset. I suspect we might be able to do better if we expanded the elf reading prior to the packaging step, but until I see that producing a concrete issue I'm reticent to go down that path.